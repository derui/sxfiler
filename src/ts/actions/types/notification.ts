// actions for notification

enum ActionTypes {
  timeout = "notification_timeout",
}

export default ActionTypes;
