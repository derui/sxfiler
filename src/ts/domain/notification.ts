
// type of notification
export enum NotificationType {
  OneShot = 'oneshot',
  Progress = 'progress',
}

export enum Level {
  Info = 'info',
  Warning = 'warning',
  Error = 'error'
}

export interface OneShotBody {
  kind: "oneshot";
  message: string;
}

export interface ProgressBody {
  kind: "progress";
  process: string;
  current: number;
  target: number;
}

export type Body = OneShotBody | ProgressBody

// notification from server
export class Notification {
  constructor(
    public readonly id: string,
    public readonly level: Level,
    public readonly body: Body
  ) {}
}
