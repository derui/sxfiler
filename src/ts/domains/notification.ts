// type of notification
export enum NotificationKind {
  Message = "message",
  Progress = "progress",
}

export enum Level {
  Info = "info",
  Warning = "warning",
  Error = "error",
}

export type MessageBody = {
  readonly message: string;
};

export type ProgressBody = {
  readonly process: string;
  readonly current: number;
  readonly target: number;
};

export type Body = MessageBody | ProgressBody;

export type MessageNotification = {
  readonly id: string;
  readonly kind: "message";
  readonly level: Level;
  readonly body: MessageBody;
};

export type ProgressNotification = {
  readonly id: string;
  readonly kind: "progress";
  readonly level: Level;
  readonly body: ProgressBody;
};

export type Notification = MessageNotification | ProgressNotification;

/**
 * factory function to create one-shot notification
 * @param id
 * @param level
 * @param message
 */
export const createMessage = (id: string, level: Level, message: string): MessageNotification => {
  return { kind: NotificationKind.Message, id, level, body: { message } };
};

/**
 * factory function to create progress notification
 * @param id
 * @param level
 * @param progress
 */
export const createProgress = (
  id: string,
  level: Level,
  progress: { process: string; current: number; target: number }
): Notification => {
  return {
    kind: NotificationKind.Progress,
    id,
    level,
    body: { ...progress },
  };
};
