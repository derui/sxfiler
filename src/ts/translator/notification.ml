open Sxfiler_rpc.Types

class type progress =
  object
    method targeted : float Js.readonly_prop

    method current : float Js.readonly_prop

    method process : Js.js_string Js.t Js.readonly_prop
  end

class type js =
  object
    method id : Js.js_string Js.t Js.readonly_prop

    method _type : Js.js_string Js.t Js.readonly_prop

    method level : int Js.readonly_prop

    method body_message : Js.js_string Js.t Js.readonly_prop

    method body_progress : progress Js.t Js.readonly_prop
  end

let of_js js : Notification.t =
  let open Sxfiler_core.Option in
  let module D = Sxfiler_domain in
  let v =
    D.Notification.Level.of_int js##.level
    >>= fun level ->
    ( match Js.to_string js##._type with
      | "message" -> Some (D.Notification.OneShot {message = Js.to_string js##.body_message})
      | "progress" ->
        Some
          (D.Notification.Progress
             { targeted = js##.body_progress##.targeted
             ; current = js##.body_progress##.current
             ; process = Js.to_string js##.body_progress##.process })
      | _ -> None )
    >>= fun body -> Some {Notification.id = Js.to_string js##.id; level; body}
  in
  get_exn v
