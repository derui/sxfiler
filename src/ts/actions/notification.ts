import { Notification } from "../domains/notification";
import { AppAction } from "./type";

export enum ActionTypes {
  timeout = "notification_timeout",
  notify = "notification_notify",
  remove = "notification_remove",
}

type TimeoutAction = AppAction<ActionTypes.timeout, { notificationId: string }>;
type NotifyAction = AppAction<ActionTypes.notify, { notification: Notification }>;
type RemoveAction = AppAction<ActionTypes.remove, { notificationId: string }>;

export type Actions = TimeoutAction | NotifyAction | RemoveAction;

const timeout = (notificationId: string): TimeoutAction => {
  return { type: ActionTypes.timeout, notificationId };
};

const notify = (notification: Notification): NotifyAction => {
  return { type: ActionTypes.notify, notification };
};

const remove = (notificationId: string): RemoveAction => {
  return { type: ActionTypes.remove, notificationId };
};

export const actions = { timeout, notify, remove };
