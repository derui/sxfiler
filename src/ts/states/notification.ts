import { createNotifications, ProgressNotifications } from "../domains/progress-notifications";

export interface State {
  // all notification not finished yet
  progresses: ProgressNotifications;
  timeouts: ProgressNotifications;
}

// get empty state
export function empty(): State {
  return {
    progresses: createNotifications([]),
    timeouts: createNotifications([]),
  };
}
