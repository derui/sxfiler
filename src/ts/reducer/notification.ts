// reducers for notification
import { Actions } from "../actions";
import ActionTypes from "../actions/types/notification";
import { empty, State } from "../types/store-state/notification";

export function reducer(state: State = empty(), action: Actions): State {
  switch (action.type) {
    case ActionTypes.timeout:
      return { ...state, notifications: state.notifications.remove(action.notificationId) };
  }
  return state;
}

export default reducer;
