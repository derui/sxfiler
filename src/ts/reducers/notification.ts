// reducers for notification
import { Actions, ActionTypes } from "../actions";
import { empty, State } from "../states/notification";

export function reducer(state: State = empty(), action: Actions): State {
  switch (action.type) {
    case ActionTypes.NOTIFICATION_REMOVE:
      return {
        ...state,
        progresses: state.progresses.remove(action.notificationId),
        timeouts: state.timeouts.remove(action.notificationId),
      };
    case ActionTypes.NOTIFICATION_TIMEOUT:
      const notification = state.progresses.findById(action.notificationId);
      if (notification) {
        return { ...state, timeouts: state.timeouts.append(notification) };
      }
      return state;
    case ActionTypes.NOTIFICATION_RECEIVE_PROGRESS:
      return { ...state, progresses: state.progresses.append(action.notification) };
  }
  return state;
}

export default reducer;
