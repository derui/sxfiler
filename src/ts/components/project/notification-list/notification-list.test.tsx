import * as React from "react";
import renderer from "react-test-renderer";

import * as N from "../../../domains/notification";
import T from "./notification-list";

describe("Project", () => {
  describe("Notification List", () => {
    it("can render correctly", () => {
      const notifications = [
        N.createOneShot("id", N.Level.Info, "message"),
      ];
      const tree = renderer.create(<T notifications={notifications} timeouts={[]} onItemTimeouted={() => { }} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("can render with empty list of notification", () => {
      const tree = renderer.create(<T notifications={[]} timeouts={[]} onItemTimeouted={() => { }} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should set timeout specified", () => {
      const notifications = [
        N.createOneShot("id", N.Level.Info, "message"),
        N.createOneShot("id2", N.Level.Warning, "message"),
      ];
      const tree = renderer.create(<T notifications={notifications} timeouts={["id"]} onItemTimeouted={() => { }} />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
