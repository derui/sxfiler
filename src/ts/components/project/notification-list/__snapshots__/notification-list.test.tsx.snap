// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Notification List can render correctly 1`] = `
<div
  className="root"
  role="list"
>
  <div
    aria-selected={false}
    className="root"
    data-level="info"
    role="listitem"
  >
    message
  </div>
</div>
`;

exports[`Project Notification List can render with empty list of notification 1`] = `
<div
  className="root"
  role="list"
/>
`;

exports[`Project Notification List should set timeout specified 1`] = `
<div
  className="root"
  role="list"
>
  <div
    aria-selected={false}
    className="root"
    data-level="warning"
    role="listitem"
  >
    message
  </div>
</div>
`;
